### DISCUSSION TITLE
How to do puppet warp / liquify mesh programmatically

### DISCUSSION THREAD

**Discussion by avisra** [ImageMagick/ImageMagick#6275] [Help]
### ImageMagick version

Latest

### Operating system, version and so on

Windows 11

### Description

We have been mimicking photoshop file functionality in ImageMagick for a set of PSDs that we have. We have been able to mimic most effects existing in the PSDs thus far. However, we recently have encountered the puppet warp and liquify mesh tool being used in some PSDs, and we don't know how to move forward with these. Puppet warp isn't exportable. And the liquify mesh tool's export isn't very useful.

Can you think of any other way to take a designer-created mesh (from a tool like photoshop, where they can visually distort the image), and import that into a CLI tool like ImageMagick to automatically process it?

**Comment by fmw42** [comment]
Imagemagick does not support those features.

**Reply by avisra** [reply]
Any ideas outside of imagemagick? Not against learning something new haha.

**Comment by fmw42** [comment]
As far as I know OpenCV does not have puppet warp either.   

If you have a puppet warp in Photoshop, you might be able to transfer its equivalent in the form of a spatial HALD 3D map to Imagemagick. But Imagemagick is not GUI enough to allow you to create and dynamically move mesh vertices around. 

I have a bash unix shell script called meshwarp for Imagemagick that allows you to manually create a mesh and render it it. See meshwarp at http://www.fmwconcepts.com/imagemagick/index.php.  You could create a GUI for it.  Check with gilles.marcellier@alioscopy.com.  He was working on something like that.

See also http://im.snibgo.com/pinorpush.htm.  User snibgo may have some other ideas, also.

**Comment by snibgo** [comment]
I don't really know what Puppet Warp is, but any displacement can be modulated by a gradient mask, so the displacement varies across the image. For example, see [Displacement by masked SRT](http://im.snibgo.com/dispmsksrt.htm).

**Comment by fmw42** [comment]
Snibgo:  Puppet warp is basically a triangle mesh for which one can move the vertices around and make them more or less dense using a GUI in Photoshop.  It then does a triangle (3-point affine, i.e. barycentric) warp on each triangle.  See https://www.youtube.com/watch?v=xRGt7byhS50. You have done something like that as I recall - that is warping a mesh of triangles.  You just do not have the GUI for manipulating the mesh.  I think it is called puppet warp, since it allows you to manipulate arms and legs in a 3D model from the mesh as if the item was a puppet.

I do not know what liquify mesh is in Photoshop. But from https://www.youtube.com/watch?v=MceH8a9JfAg, it seems to be an interactive ability to apply local distortions to a rectangular mesh to distort the image under the mesh.  I think you may have do some similar things that you show on your web page, though I think your grid is just for show. But you can give a proper description of what you have done in case I am wrong about it.

**Comment by snibgo** [comment]
Thanks, fmw42. Yes, the meshes on my pages are just to show more clearly what transformations have occurred.

IM doesn't have Delauney Triangulation. If it did, we could give a series of points on the image. IM could then figure out the triangles, and distort each individual triangle. The effect would be like your `meshwarp`, but faster and without needing to specifying each triangle.

The PS Puppet Mesh takes the automation further, as it first defines locations, and then creates the triangles. The video you linked also shows options for locking aspects of the triangulated geometry.

The PS Liquify Mesh video seems close to my [Displacement by masked SRT](http://im.snibgo.com/dispmsksrt.htm). In the video the user defines an operation such as shrink or enlarge, and the mask is a radial gradient so the effect is at a maximum at a point, with decreasing effect further from that point. In the video, the mesh seems irrelevant, and just for show.

**Comment by avisra** [comment]
I've been reading through all of the comments, and admittedly - I am out of my realm expertise. If there is not a good replacement for Photoshop's puppet warp programmatically, then is there something within ImageMagick that could achieve the same requirement?

Assume I need to place a full-art design onto this blanket in this image: https://www.bhg.com/thmb/hejbs00dXRnrFfMjH7BrOs2DS_o=/fit-in/1002x1002/filters:no_upscale():strip_icc()/cozyearth-4.2020_0433_73f88f97-1f18-47bc-898c-c22def43bd4c_1200x1200-3db60dc33b5045928279595ac07e8ef4.jpg

I can figure out the lighting, the general bump displacement (using a greyscale displacement map), etc within ImageMagick. But the part that I'm getting hung up on is being able to change the direction of the design as it goes along the curvature of the chair. Is this doable with a displacement map?

**Comment by snibgo** [comment]
> But the part that I'm getting hung up on is being able to change the direction of the design as it goes along the curvature of the chair. Is this doable with a displacement map?

Yes.

Suppose we want to "paint" this image, `toes_grid.jpg`. The grid lines are only so we can see what is happening.

![toes_grid](https://user-images.githubusercontent.com/33812368/235047577-4b9a5b60-7ade-435d-9cd7-69285e4ebe63.jpg)

I have named your linked image `blanket.jpg`. First, we crop to the approximate area where we will paint the distorted image. We also ensure that no colour channel is at the maximum. Windows syntax:
```
magick ^
  blanket.jpg ^
  -crop 380x320+412+188 +repage ^
  -channel R -fx "u==1?0.99:u" ^
  -channel G -fx "u==1?0.99:u" ^
  -channel B -fx "u==1?0.99:u" ^
  +channel ^
  b2.png
```
We edit the file `b2.png` with Gimp "pencil", painting solid blue lines to mark the top and bottom boundary of where we want the distortion, and green lines for the left and right sides. These lines must extend to the edges of the image. The line thicknesses don't matter. The distorted image will be just inside the marked lines. The result of this edit is `b2e.png`:

![b2e](https://user-images.githubusercontent.com/33812368/235047598-0ac8b0b1-fa25-4349-a9dc-6346d5cff332.png)


Then we run this IM command (Windows BAT syntax):
```
magick ^
  b2e.png ^
  -define compose:clamp=on ^
  -define quantum:format=floating-point ^
  -depth 32 ^
  -write mpr:INP ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  -set option:WW %%w ^
  -set option:HH %%h ^
  -channel B -separate +channel ^
  -fill Black +opaque White ^
  -fill White ^
  -draw "color 0,0 floodfill" ^
  -draw "color 0,%%[fx:h-1] floodfill" ^
  ( -clone 0 ^
    -alpha off ^
    -negate ^
    -write mpr:MASKVERT +delete ^
  ) ^
  ( -clone 0 ^
    -fill Black ^
    -draw "color 0,0 floodfill" ^
    -channel RGB -negate +channel ^
    -scale "%%[WW]x1^!" ^
    -size 1x%%[HH] gradient:Black-White ^
    -scale "%%[WW]x%%[HH]^!" ^
    -compose DivideDst -composite ^
  ) ^
  ( -clone 0 ^
    -fill Black ^
    -draw "color 0,%%[fx:h-1] floodfill" ^
    -channel RGB -negate +channel ^
    -scale "%%[WW]x1^!" ^
    -size 1x%%[HH] gradient:White-Black ^
    -scale "%%[WW]x%%[HH]^!" ^
    -compose DivideDst -composite ^
    -channel RGB -negate +channel ^
  ) ^
  -delete 0 ^
  -fx "v/(v-u+1)" ^
  -write mpr:DVERT +delete ^
  mpr:INP ^
  -channel G -separate +channel ^
  -fill Black +opaque White ^
  -fill White ^
  -draw "color 0,0 floodfill" ^
  -draw "color %%[fx:w-1],0 floodfill" ^
  ( -clone 0 ^
    -alpha off ^
    -negate ^
    -write mpr:MASKHORZ +delete ^
  ) ^
  ( -clone 0 ^
    -fill Black ^
    -draw "color 0,0 floodfill" ^
    -channel RGB -negate +channel ^
    -scale "1x%%[HH]^!" ^
    -size %%[WW]x1 gradient:Black-White ^
    -scale "%%[WW]x%%[HH]^!" ^
    -compose DivideDst -composite ^
  ) ^
  ( -clone 0 ^
    -fill Black ^
    -draw "color %%[fx:w-1],0 floodfill" ^
    -channel RGB -negate +channel ^
    -scale "1x%%[HH]^!" ^
    -size %%[WW]x1 gradient:White-Black ^
    -scale "%%[WW]x%%[HH]^!" ^
    -compose DivideDst -composite ^
    -channel RGB -negate +channel ^
  ) ^
  -delete 0 ^
  -fx "v/(v-u+1)" ^
  -write mpr:DHORZ ^
  mpr:DVERT ^
  -size %%[WW]x%%[HH] xc:Black ^
  -combine ^
  ( toes_grid.jpg -resize "%%[WW]x%%[HH]^!" ) ^
  +swap ^
  -compose Distort -composite ^
  ( mpr:MASKVERT mpr:MASKHORZ ^
    -alpha off ^
    -compose Darken -composite ^
  ) ^
  -compose CopyOpacity -composite ^
  blanket.jpg ^
  +swap ^
  -geometry +412+188 ^
  -compose Over -composite ^
  c.png
```
Here is the result, `c.png`. The arm of the chair has been over-painted by the distorted image. That could be fixed using Gimp.

![c](https://user-images.githubusercontent.com/33812368/235047621-1d49d8f7-5b02-4ea9-8a7a-0f4aff9ea771.jpg)


The command is long, but not very complicated. What does it do?

It makes an absolute distortion map, at the `-combine` near the bottom. That combines images for the red, green and blue channels. The blue channel is entirely zero ("black"). The red channel will vary from 0.0 to 1.0 across where we want the distortion. The green channel will vary from 0.0 to 1.0 from the top of where we want the distortion to the bottom.

The command makes the green channel first. It does this by creating two gradients: the first varies from 0.0 at the top of the image to 1.0 at the bottom of where we want the distortion, and second is 0.0 at the top of where we want the distortion to 1.0 at the bottom of the image. It combines the two gradients with `-fx "v/(v-u+1)"`, an operation that I call a "fan composition".

So that is the green channel. The red channel is the same, but from left to right. The command combines the red, green and blue channels to form the distortion map.

Then it reads `toes_grid.jpg` and distorts it with the map.

Then it makes pixels outside the distorted image transparent.
